### ISSUE TITLE
Metrictank OOM due to expensive queries

### ISSUE THREAD

**Issue by shanson7** [grafana/metrictank#968]
Right now, metrictank can short-circuit render requests based on the number of datapoints it think it would have to fetch. This is good. What it doesn't do is try to short-circuit a fetch to get the series it needs based on the number of series that could be returned from such an expression. 

We saw issues when trying to pull back a request for 8 million time-series where the amount of data killed metrictank before it even was able to get to the fact that it wouldn't be able to handle the request due to the number of datapoints. Metrictank allocated 12GB of memory in 30 seconds and then got killed.

**Comment by shanson7** [comment]
I think that this could be solved fairly easy with a few small changes:

1. peerQuery[Speculative] should load up a channel of responses rather than gathering them and then processing them all at once. This would allow us to abort requests earlier before potential crashing.
2. Try to reuse buffers as we process requests. This is just an optional optimization.
3. Add a parameter like `max-render-series` similar to the `max-points-per-req-hard`.

**Comment by Dieterbe** [comment]
dieter [11:57 AM]
do i understand correctly that the problem is basically memory exhaustion due to to many series, but we don't know there's too many series, until we have queried the index, and this query causes the exhaustion in the first place (either bytag or without tags)


shanson [11:59 AM]
yeah
The memory footprint of the series themselves is larger than that of datapoints. So a query with 10 series at a millions datapoints each is significantly different than one with 1 million series with 10 datapoints each (edited)

dieter [12:02 PM]
it surprises me that a Find() call could exhaust memory, considering it queries a fraction of what's in the index, and the index is all in memory

shanson [12:02 PM]
Oh, no. This is on the `render` node
So, in our 120 shard group cluster each instance has only ~1/120th of the metadata (edited)
but a poorly designed query could ask for all of it to funnel back to a single instance

dieter [12:04 PM]
ooh...
DDOS attack

shanson [12:05 PM]
DDOS’d by his own brethren :slightly_smiling_face:

dieter [12:07 PM]
so basically speculativePeerQuery provides a streaming api instead of a blocking-wait-for-everything api towards it caller
for certain callers that's useful, for others not so much

shanson [12:07 PM]
Yeah, I also provided a wrapper around the channel one that does the current map building thing
because some responses are fixed size and don’t need the added complexity

dieter [12:08 PM]
sounds neat
sounds :thumbsup:
